Code Templates

jvelilla opened the pull request:
Code templates mechanism for EiffelStudio.
- Definition using Eiffel code.

jvelilla committed:
Added Code Templates: definition using Eiffel code.
Added a service to get the catalog of templates.
Integrated as part of code completion.

jvelilla committed:
Merge branch 'master' of https://github.com/EiffelSoftware/EiffelStudio into es_template

manuseiffel commented on Src/Delivery/studio/templates/code/array_template.e:
```diff
@@ -0,0 +1,32 @@
+class ARRAY_TEMPLATE [T -> COMPARABLE]
```
Not required but I think it would be nice to have a note clause at the top with our usual tags: description, license, date, revision.

jvelilla replied:
Ok, I will add it.

manuseiffel commented on Src/Eiffel/interface/new_graphical/text_window/interface/eb_smart_editor.e:
```diff
@@ -117,7 +118,7 @@ feature {NONE} -- Access
 				Result := l_cookie
 			else
-				Result := (create {UUID_GENERATOR}).generate_uuid
+				Result := (create {UUID_GENERATOR}).generate_uuid.as_attached
```
Drop the as_attached.

jvelilla replied:
ok

jvelilla committed:
Added top notes to the code templates.
Updated code template definition class to process top notes.
Removed the as_attached.

jvelilla committed:
Update code template:
Added a new class providing most common constants used in code templates.
Fixed issues with templates as commands.

jvelilla committed:
Updated code template to accept queries and commands with one argument.
Removed unneeded code.

jvelilla committed:
Updated code with a collection of tokens (linked) to be used by
SMART_TEXT.

jocelyn committed:
Merge branch 'master' into code_template

jocelyn committed:
Cleaned unused local variables, and removed an assignment attempt.

jvelilla committed:
Added: input arguments and default values for them.
Default values are expressed as STRINGs and the order define how to initialize
them. f (a,b,c) and a note default: a.lower, a.upper means b = a.lower and c = a.upper
The first argument define the context.

jvelilla committed:
Merge branch 'code_template' of https://github.com/jocelyn/EiffelStudio into es_template

jvelilla committed:
Updated code to handle renaming with arguments and defaults input values.

jvelilla committed:
Updated template using the new definition.

jocelyn committed:
Fixed infinite loop.

jocelyn committed:
Fixed calls on void target (related to read_only_locals).

jvelilla committed:
Added global templates to the list of completion possiblities.
Added tooltip to show the previw of the code template.
Fixed code completion issues.

jvelilla committed:
Initial implementation of filter templates to show in the completion possibilities dialog.
Refactor the the code template initialization to enable descendants to redefine
how to populate the list of sorted names.
Fixed issue locals name generation in EB_TEMPLATE_FOR_COMPLETION.

jvelilla committed:
Merge branch 'code_template' of https://github.com/jocelyn/EiffelStudio into es_template

jvelilla committed:
Added missing class

jocelyn committed:
Updated linked editing to allow severals linked editing at the same time.
Used it for code template inline variable editing (only from the first token of the linked collection).
Fixed issue with empty completion list.

jvelilla committed:
Added default input values as part of linked tokens.

jvelilla committed:
Updated code to show code templates in the tooltip with code highlithed.

jvelilla committed:
Implemented Crlt+Space cylce (features, templates, features) when we display
completion possibilities.
Added a new option bar for templates.

jvelilla commented:
Initial import of code template on revision 99107
